### DISCUSSION TITLE
[java] Cache invalidation - auxClasspath or classpath?

### DISCUSSION THREAD

**Discussion by fflatorre** [pmd/pmd#3040] [Q&A] [answered]
Please tag this issue as a question.

Hi,

I'm trying to gt my head around on when the cache gets invalidated.

In the official documentation, among the other reasons, it's mentioned teh cache is invalidated for a change in the auxclasspath :
https://pmd.github.io/pmd-6.0.0/pmd_release_notes_old.html#22-april-2017---560

We are not setting the "auxclasspath" as mentioned in the guide here :
https://pmd.github.io/pmd-6.2.0/pmd_userdocs_tools_ant.html

However, looking at the code :
https://github.com/pmd/pmd/blob/71df067a0dbd27ef78e4c82ca7cc58fd138f65c3/pmd-core/src/main/java/net/sourceforge/pmd/cache/FileAnalysisCache.java

it looks like that is the classpath instead that gets check-summed and then persisted/retrieved from the cache file.

Can anyone please clarify this aspect ?

Cheers,
Francesco

**Affects PMD Version:** 
6.2.0
**Rule:**
NA
**Description:**
NA
**Code Sample demonstrating the issue:**

```
NA
```

**Running PMD through:** *[Ant]*

**Comment by jsotuyod** [comment, marked as answer]
@fflatorre over the last few releases the feature has matured a lot. Currently the cache is completely invalidated if

1. The ruleset changed
https://github.com/pmd/pmd/blob/798c027d8d8b05843d0909b42861ba5f0bdd829c/pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java#L107
2. The auxclasspath changed
https://github.com/pmd/pmd/blob/798c027d8d8b05843d0909b42861ba5f0bdd829c/pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java#L120
3. The execution classpath chamged
https://github.com/pmd/pmd/blob/798c027d8d8b05843d0909b42861ba5f0bdd829c/pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java#L132

So, under current releases, both the auxclasspath AND the java classpath invalidates the cache. Moreover, the old code just checked the value of the auxclasspath (file1.jar:file2.jar), but current releases actually check both the order and the contents.

Let me know if you have any questions

**Comment by fflatorre** [comment]
Cheers @jsotuyod I got lost between releases.
So pull request [#687](https://github.com/pmd/pmd/pull/687 ) resolves issue [#603](https://github.com/pmd/pmd/issues/603) released on [6.1.0](https://pmd.github.io/pmd-6.1.0/pmd_release_notes_old.html#incremental-analysis) which is the bit I was missing.

Great job !
Francesco
